ta/pkcs11/src/serializer.c

```diff
@@ -24,58 +26,72 @@ void serialargs_init(struct serialargs *args, void *in, size_t size)
 
 enum pkcs11_rc serialargs_get(struct serialargs *args, void *out, size_t size)
 {
-	if (args->next + size > args->start + args->size) {
-		EMSG("arg too short: full %zd, remain %zd, expect %zd",
-		     args->size, args->size - (args->next - args->start), size);
-		return PKCS11_CKR_ARGUMENTS_BAD;
-	}
+	enum pkcs11_rc rc = PKCS11_CKR_OK;
+	void *src = NULL;
 
-	TEE_MemMove(out, args->next, size);
-
-	args->next += size;
+	rc = serialargs_get_ptr(args, &src, size);
+	if (!rc)
+		TEE_MemMove(out, src, size);
 
 	return PKCS11_CKR_OK;
 }
 
-enum pkcs11_rc serialargs_alloc_and_get(struct serialargs *args,
-					void **out, size_t size)
+static enum pkcs11_rc alloc_and_get(struct serialargs *args, char *orig_next,
+				    const void *buf0, size_t buf0_sz,
+				    void **out, size_t size)
 {
-	void *ptr = NULL;
+	enum pkcs11_rc rc = PKCS11_CKR_OK;
+	uint8_t *ptr = NULL;
+	void *src = NULL;
+	size_t sz = 0;
 
-	if (!size) {
+	if (ADD_OVERFLOW(buf0_sz, size, &sz))
+		return PKCS11_CKR_ARGUMENTS_BAD;
+
+	if (!sz) {
 		*out = NULL;
 		return PKCS11_CKR_OK;
 	}
 
-	if (args->next + size > args->start + args->size) {
-		EMSG("arg too short: full %zd, remain %zd, expect %zd",
-		     args->size, args->size - (args->next - args->start), size);
-		return PKCS11_CKR_ARGUMENTS_BAD;
-	}
+	rc = serialargs_get_ptr(args, &src, size);
+	if (rc)
+		return rc;
 
-	ptr = TEE_Malloc(size, TEE_MALLOC_FILL_ZERO);
-	if (!ptr)
+	ptr = TEE_Malloc(sz, TEE_MALLOC_FILL_ZERO);
+	if (!ptr) {
+		args->next = orig_next;
 		return PKCS11_CKR_DEVICE_MEMORY;
+	}
 
-	TEE_MemMove(ptr, args->next, size);
+	TEE_MemMove(ptr, buf0, buf0_sz);
+	TEE_MemMove(ptr + buf0_sz, src, size);
 
-	args->next += size;
 	*out = ptr;
 
 	return PKCS11_CKR_OK;
 }
 
+enum pkcs11_rc serialargs_alloc_and_get(struct serialargs *args,
```

Comment:
This function is never used but exposed through the header. Why?

Reply:
It's part of the package, I think it will be used in later patches.

Reply:
Ok then. But in general PRs should be kept to a minimal scope. Reviewing this 3000+ LOC PR is no fun. :)

Reply:
Agree, thanks for the effort.